src/buildHTML.js

```diff
@@ -88,146 +64,174 @@ export const buildExpression = function(
+        const partialGroup = checkPartialGroup(node);
+        if (partialGroup) {
+            traverseNodes(partialGroup.children, callback, prev); // Recursive DFS
+            continue;
+        }
 
-    const groups = [];
-    let j = 0;
-    for (let i = 0; i < rawGroups.length; i++) {
-        groups.push(rawGroups[i]);
-
-        // For any group that is not a space, get the next non-space.  Then
-        // lookup what implicit space should be placed between those atoms and
-        // add it to groups.
-        if (rawGroups[i].classes[0] !== "mspace" && j < nonSpaces.length - 1) {
-            // if current non-space node is left dummy span, add a glue before
-            // first real non-space node
-            if (j === 0) {
-                groups.pop();
-                i--;
-            }
+        // Ignore explicit spaces (e.g., \;, \,) when determining what implicit
+        // spacing should go between atoms of different classes
+        if (node.classes[0] === "mspace") {
+            continue;
+        }
 
-            // Get the type of the current non-space node.  If it's a document
-            // fragment, get the type of the rightmost node in the fragment.
-            const left = getTypeOfDomTree(nonSpaces[j], "right");
-
-            // Get the type of the next non-space node.  If it's a document
-            // fragment, get the type of the leftmost node in the fragment.
-            const right = getTypeOfDomTree(nonSpaces[j + 1], "left");
-
-            // We use buildExpression inside of sizingGroup, but it returns a
-            // document fragment of elements.  sizingGroup sets `isRealGroup`
-            // to false to avoid processing spans multiple times.
-            if (left && right && isRealGroup) {
-                const nonSpacesJp1: HtmlDomNode = assert(nonSpaces[j + 1]);
-                const space = isLeftTight(nonSpacesJp1)
-                    ? tightSpacings[left][right]
-                    : spacings[left][right];
-
-                if (space) {
-                    let glueOptions = options;
-
-                    if (expression.length === 1) {
-                        const node =
-                            checkNodeType(expression[0], "sizing") ||
-                            checkNodeType(expression[0], "styling");
-                        if (!node) {
-                            // No match.
-                        } else if (node.type === "sizing") {
-                            glueOptions = options.havingSize(node.size);
-                        } else if (node.type === "styling") {
-                            glueOptions = options.havingStyle(styleMap[node.style]);
-                        }
-                    }
-
-                    groups.push(buildCommon.makeGlue(space, glueOptions));
-                }
+        const result = callback(node, prev.node);
+        if (result) {
+            if (prev.insertAfter) {
+                prev.insertAfter(result);
```

Comment:
I've tried inserting before the current node, but it inserts spacing after the`\newline` and probably breaks other things.

Reply:
Isn't that what the else branch does?

Reply:
Could you explain the `\newline` situation a bit more?

Reply:
That makes sense, thanks.